Configuration cache error with Java11 collections

### Current Behavior

When using Java11 collections Gradle fails with configuration cache with error:
```
org.gradle.api.GradleException: Could not load the value of field `$myList` of `Build_gradle$$$result$1$1` bean found in field `action` of `org.gradle.api.internal.AbstractTask$TaskActionWrapper` bean found in field `actions` of task `:testJava11Collections` of type `org.gradle.api.DefaultTask`.
        at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:112)
        at org.gradle.configurationcache.serialization.beans.BeanPropertyReader.readStateOf(BeanPropertyReader.kt:68)
        at org.gradle.configurationcache.serialization.codecs.BeanCodec.readBeanOf(BeanCodec.kt:72)
        at org.gradle.configurationcache.serialization.codecs.BeanCodec.decode(BeanCodec.kt:47)
        at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1$decodeLoop$1.invokeSuspend(Combinators.kt:166)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.configurationcache.serialization.codecs.jos.JavaObjectSerializationCodec.readResolve(JavaObjectSerializationCodec.kt:194)
        at org.gradle.configurationcache.serialization.codecs.jos.JavaObjectSerializationCodec.decode(JavaObjectSerializationCodec.kt:109)
        at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
        at org.gradle.configurationcache.serialization.DefaultReadContext.read(Contexts.kt:269)
        at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:106)
        ... 169 more
Caused by: java.io.InvalidObjectException: null array
        ... 177 more
        ````

### Expected Behavior

There is no error, Gradle works with Java11 collections.

### Context (optional)

Cannot use a plugin that uses Java11 collections.

### Steps to Reproduce

1. git clone git@github.com:asodja/gradle-cc-with-java11-collections.git
2. Use Java11 to run a project
2. Run `./gradlew :testPreJava11Collections --configuration-cache --stacktrace` -> everything works
3. Run `./gradlew :testJava11Collections --configuration-cache --stacktrace` -> exception is thrown


### Gradle version

Any version, reproducer uses `8.6-20231104002619+0000` nightly 

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheJavaCollectionsIntegrationTest.groovy
+++ b/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheJavaCollectionsIntegrationTest.groovy
@@ -1,0 +1,56 @@
+/*
+ * Copyright 2025 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl
+
+
+import org.gradle.test.precondition.Requires
+import org.gradle.test.preconditions.UnitTestPreconditions
+import spock.lang.Issue
+
+class ConfigurationCacheJavaCollectionsIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    @Issue('https://github.com/gradle/gradle/issues/26942')
+    @Requires(value = UnitTestPreconditions.Jdk11OrLater)
+    def 'restores Java 11 collections'() {
+        given:
+        buildFile '''
+            import java.util.*;
+
+            class CollectionsTask extends DefaultTask {
+                private List<String> list = List.of('foo', 'bar')
+                private Set<String> set = Set.of('foo', 'bar')
+                private Map<String, String> map = Map.of('foo', 'bar')
+
+                @TaskAction def print() {
+                    [list: list, set: set.sort(), map: map].each {
+                        println it
+                    }
+                }
+            }
+
+            tasks.register('collections', CollectionsTask)
+        '''
+
+        when:
+        configurationCacheRun 'collections'
+
+        then:
+        outputContains 'list=[foo, bar]'
+        outputContains 'set=[bar, foo]'
+        outputContains 'map={foo=bar}'
+    }
+}

--- a/platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaObjectSerializationCodec.kt
+++ b/platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaObjectSerializationCodec.kt
@@ -175,7 +175,7 @@ class JavaObjectSerializationCodec(
 
                     replacement::class.java === value::class.java -> {
                         // Avoid a StackOverflowException when the replacement and value are of the same type.
-                        // TODO:configuration-cache This is likely incorrect when the class also supports the `writeObject` protocol.
+                        // TODO:configuration-cache Skipping Java serialization for the replacement is likely incorrect when the class also supports the `writeObject` protocol
                         writeEnum(Format.ReadResolveBean)
                         encodeBean(value)
                     }

--- a/platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaObjectSerializationCodec.kt
+++ b/platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaObjectSerializationCodec.kt
@@ -30,6 +30,7 @@ import org.gradle.internal.serialize.graph.decodePreservingIdentity
 import org.gradle.internal.serialize.graph.encodeBean
 import org.gradle.internal.serialize.graph.encodePreservingIdentityOf
 import org.gradle.internal.serialize.graph.readEnum
+import org.gradle.internal.serialize.graph.readNonNull
 import org.gradle.internal.serialize.graph.withBeanTrace
 import org.gradle.internal.serialize.graph.withImmediateMode
 import org.gradle.internal.serialize.graph.writeEnum
@@ -107,8 +108,13 @@ class JavaObjectSerializationCodec(
                     }
                 }
 
-                Format.ReadResolve -> {
+                Format.ReadResolveBean -> {
                     readResolve(decodeBean())
+                        .also { putIdentity(id, it) }
+                }
+
+                Format.ReadResolveAny -> {
+                    readResolve(readNonNull())
                         .also { putIdentity(id, it) }
                 }
 
@@ -159,14 +165,25 @@ class JavaObjectSerializationCodec(
         override suspend fun WriteContext.encode(value: Any) {
             encodePreservingIdentityOf(value) {
                 val replacement = writeReplace.invoke(value)
-                if (replacement is SerializedLambda) {
-                    writeEnum(Format.SerializedLambda)
-                    SerializedLambdaParametersCheckingCodec.run {
-                        encode(replacement)
+                when {
+                    replacement is SerializedLambda -> {
+                        writeEnum(Format.SerializedLambda)
+                        SerializedLambdaParametersCheckingCodec.run {
+                            encode(replacement)
+                        }
                     }
-                } else {
-                    writeEnum(Format.ReadResolve)
-                    encodeBean(replacement)
+
+                    replacement::class.java === value::class.java -> {
+                        // Avoid a StackOverflowException when the replacement and value are of the same type.
+                        // TODO:configuration-cache This is likely incorrect when the class also supports the `writeObject` protocol.
+                        writeEnum(Format.ReadResolveBean)
+                        encodeBean(value)
+                    }
+
+                    else -> {
+                        writeEnum(Format.ReadResolveAny)
+                        write(replacement)
+                    }
                 }
             }
         }
@@ -176,7 +193,7 @@ class JavaObjectSerializationCodec(
     object ReadResolveEncoding : Encoding {
         override suspend fun WriteContext.encode(value: Any) {
             encodePreservingIdentityOf(value) {
-                writeEnum(Format.ReadResolve)
+                writeEnum(Format.ReadResolveBean)
                 encodeBean(value)
             }
         }
@@ -184,7 +201,8 @@ class JavaObjectSerializationCodec(
 
     private
     enum class Format {
-        ReadResolve,
+        ReadResolveBean,
+        ReadResolveAny,
         WriteObject,
         ReadObject,
         SerializedLambda

--- a/platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaSerializationEncodingLookup.kt
+++ b/platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/jos/JavaSerializationEncodingLookup.kt
@@ -45,8 +45,8 @@ class JavaSerializationEncodingLookup {
     fun calculateEncoding(type: Class<*>): EncodingDetails {
         val candidates = type.allMethods()
         val encoding = writeReplaceEncodingFor(candidates)
-            ?: readResolveEncodingFor(candidates)
             ?: writeObjectEncodingFor(candidates)
+            ?: readResolveEncodingFor(candidates)
             ?: readObjectEncodingFor(candidates)
         return EncodingDetails(encoding)
     }

